demo/lightning/checkpoint/simulated/llama2_utils.py

```diff
@@ -0,0 +1,161 @@
+        (hidden_dim, hidden_dim), empty)
+    state_dict[f'{prefix}attn.wv.weight'] = create_tensor(
+        (hidden_dim, hidden_dim), empty)
+    state_dict[f'{prefix}attn.wo.weight'] = create_tensor(
+        (hidden_dim, hidden_dim), empty)
+
+
+def add_ffn_weights(state_dict: Dict[str, torch.Tensor], prefix: str,
+                    hidden_dim: int, intermediate_dim: int,
+                    empty: bool) -> None:
+    """Adds feed-forward network weights directly to state dict."""
+    state_dict[f'{prefix}ff.w1.weight'] = create_tensor(
+        (intermediate_dim, hidden_dim), empty)
+    state_dict[f'{prefix}ff.w2.weight'] = create_tensor(
+        (hidden_dim, intermediate_dim), empty)
+    state_dict[f'{prefix}ff.w3.weight'] = create_tensor(
+        (intermediate_dim, hidden_dim), empty)
+
+
+def add_adamw_states(state_dict: Dict[str, torch.Tensor], prefix: str,
+                     param_prefix: str, shape: tuple, empty: bool) -> None:
+    """Adds AdamW optimizer states directly to state dict."""
+    state_dict[f'{prefix}{param_prefix}.exp_avg'] = create_tensor(shape, empty)
+    state_dict[f'{prefix}{param_prefix}.exp_avg_sq'] = create_tensor(
+        shape, empty)
+
+
+def add_layer_weights(state_dict: Dict[str, torch.Tensor], prefix: str,
+                      hidden_dim: int, intermediate_dim: int, empty: bool,
+                      use_adamw: bool) -> None:
+    """Adds all weights for a single transformer layer directly to state dict."""
+    # Attention weights
+    add_attention_weights(state_dict, prefix, hidden_dim, empty)
+
+    # Attention norm
+    state_dict[f'{prefix}attn_norm.weight'] = create_tensor((hidden_dim, ),
+                                                            empty)
+
+    # FFN weights
+    add_ffn_weights(state_dict, prefix, hidden_dim, intermediate_dim, empty)
+
+    # FFN norm
+    state_dict[f'{prefix}ffn_norm.weight'] = create_tensor((hidden_dim, ),
+                                                           empty)
+
+    # Add AdamW states if needed
+    if use_adamw:
+        # Attention AdamW states
+        for weight_name in ['wq', 'wk', 'wv', 'wo']:
+            add_adamw_states(state_dict, prefix, f'attn.{weight_name}',
+                             (hidden_dim, hidden_dim), empty)
+
+        # Attention norm AdamW states
+        add_adamw_states(state_dict, prefix, 'attn_norm', (hidden_dim, ),
+                         empty)
+
+        # FFN AdamW states
+        add_adamw_states(state_dict, prefix, 'ff.w1',
+                         (intermediate_dim, hidden_dim), empty)
+        add_adamw_states(state_dict, prefix, 'ff.w2',
+                         (hidden_dim, intermediate_dim), empty)
+        add_adamw_states(state_dict, prefix, 'ff.w3',
+                         (intermediate_dim, hidden_dim), empty)
+
+        # FFN norm AdamW states
+        add_adamw_states(state_dict, prefix, 'ffn_norm', (hidden_dim, ), empty)
+
+
+def create_llama2_state_dict(world_size: int,
```

Comment:
We may want to specify in a comment or in function name that this is generating distributed weights according to the FSDP style split, where each node does only gets a `1/world_size` volume of data.

Reply:
Done